File: src/omni/api/system.py
"""System status: loop freshness and fill health from the data itself.

The scheduler and autonomous runner have no heartbeat table -- and they should
not need one. A loop that is alive writes rows; a loop that is dead stops. So
the cheapest, most honest health signal is the timestamp of each loop's last
output, read straight from the tables it writes. A prediction loop that last
wrote an hour ago is either idle or stuck; this endpoint surfaces the age so a
caller can tell.

The ``loops`` array is that effect-derived view. The ``health`` object is the
process view, read from ``loop_health`` (written by the scheduler each
iteration): it carries the two signals the effect view cannot give -- a
chronically-failing loop's ``consecutive_failures`` and ``last_error`` (an
iterating-but-failing loop stops writing output, so its age looks like idle),
and an alive-but-idle loop's fresh ``last_success_at`` (sweep finding no gaps
writes nothing, yet is healthy). ``health.overall`` is the worst per-loop
verdict so a glance is enough: ok / stale / failing.

Read-only and operator-only. It reveals loop freshness, demand volume and
provider fill rates. The public ``/health`` stays minimal for anonymous uptime
checks.
"""

from __future__ import annotations

from datetime import UTC, datetime

from neutron import App, Router
from neutron.error import forbidden, unauthorized
from starlette.requests import Request

from omni.auth import resolve_audience_from_request, resolve_role_from_request
from omni.scheduler.health import EXPECTED_OPERATION_INTERVALS

# A loop is stale when its last success is older than this many times its own
# scheduled interval -- a loop missing several consecutive beats is stuck, not
# idle. The multiplier is generous on purpose: a healthy loop that takes an
# occasional slow cycle should not read as stale.
_STALE_INTERVAL_MULTIPLIER = 5.0

# Severity ordering for the overall verdict. Higher == worse.
_SEVERITY = {"ok": 0, "stale": 1, "failing": 2}



def build_router(app: App) -> Router:
    router = Router()

    @router.get("/system/status")
    async def system_status(request: Request) -> dict:
        if resolve_audience_from_request(request) is None:
            raise unauthorized("Authentication required")
        if resolve_role_from_request(request) != "operator":
            raise forbidden("Operator access required")
        pool = app.db.pool
        now = datetime.now(UTC)

        # Each loop's last output. NULL means the loop has never run (a fresh
        # deployment), which is a distinct state from stale.
        freshness = await pool.fetch(
            """
            SELECT 'prediction' AS loop_name, MAX(created_at) AS last_at FROM prediction
            UNION ALL SELECT 'finding', MAX(created_at) FROM finding WHERE status = 'surfaced'
            UNION ALL SELECT 'fill', MAX(finished_at) FROM fill_attempt
            UNION ALL SELECT 'demand', MAX(created_at) FROM demand
            UNION ALL SELECT 'claim_ingest', MAX(observed_at) FROM claim
            """
        )
        loops = []
        for row in freshness:
            last_at = row["last_at"]
            age_s = (now - last_at).total_seconds() if last_at is not None else None
            loops.append(
                {
                    "loop": row["loop_name"],
                    "last_activity": last_at.isoformat() if last_at else None,
                    "age_seconds": age_s,
                    "never_run": last_at is None,
                }
            )

        # Demand state: how much outstanding work is there, and how much has the
        # system given up on (unfillable) recently?
        demand = await pool.fetchrow(
            """
            SELECT
              count(*) FILTER (WHERE active) AS active_demand,
              count(*) AS total_demand
            FROM demand
            """
        )

        # Fill outcomes in the last hour: filled vs unfillable vs error. A rising
        # error count is a provider failing (auth, network, schema); a rising
        # unfillable count is demand the catalog cannot satisfy.
        fills = await pool.fetch(
            """
            SELECT outcome, count(*) AS n
            FROM fill_attempt
            WHERE finished_at > now() - interval '1 hour'
            GROUP BY outcome
            """
        )
        fill_recent = {row["outcome"]: row["n"] for row in fills}

        # Production in the last 24h: the throughput that proves the engine is
        # turning, not just idling.
        production = await pool.fetchrow(
            """
            SELECT
              count(*) FILTER (WHERE created_at > now() - interval '24 hours') AS predictions_24h,
              (SELECT count(*) FROM finding WHERE created_at > now() - interval '24 hours') AS findings_24h
            FROM prediction
            """
        )

        # Claim-store throughput: the volume of data under management and the
        # rate it is arriving. The board answers "what is being processed";
        # a store that stops growing while demand stays active is a provider
        # outage wearing a healthy face.
        claims = await pool.fetchrow(
            """
            SELECT
              count(*) AS total,
              count(*) FILTER (WHERE observed_at > now() - interval '24 hours') AS recent
            FROM claim
            """
        )

        # Process health per loop, from the state rows the scheduler writes each
        # iteration. Complements the effect-derived `loops` above: this is where
        # a chronically-failing or never-yet-succeeded loop becomes visible, and
        # where an alive-but-idle loop reads healthy rather than ambiguously
        # aged. No rows (a fresh deployment) is honest emptiness, not "ok" -- the
        # effect-derived loops already report never_run for that case.
        health_rows = await pool.fetch(
            """
            SELECT loop_name, last_success_at, last_failure_at,
                   consecutive_failures, last_error, expected_interval_seconds,
                   last_status, last_result
            FROM loop_health
            """
        )
        health_loops = []
        worst_severity = -1
        worst_state: str | None = None
        recorded = {r["loop_name"]: r for r in health_rows}
        operation_names = [
            *EXPECTED_OPERATION_INTERVALS,
            *sorted(set(recorded) - set(EXPECTED_OPERATION_INTERVALS)),
        ]
        for operation_name in operation_names:
            r = recorded.get(operation_name)
            default_interval = EXPECTED_OPERATION_INTERVALS.get(operation_name)
            if r is None:
                health_loops.append(
                    {
                        "loop": operation_name,
                        "state": "never_run",
                        "last_status": None,
                        "last_success_at": None,
                        "last_failure_at": None,
                        "consecutive_failures": 0,
                        "last_error": None,
                        "last_result": None,
                        "expected_interval_seconds": default_interval,
                    }
                )
                continue
            consec = int(r["consecutive_failures"])
            last_success = r["last_success_at"]
            interval = r["expected_interval_seconds"] or default_interval
            # A loop with open failures is failing even if it succeeded recently
            # -- the current iteration raised. last_success NULL means it has
            # only ever failed, which is failing too.
            if consec > 0 or r["last_status"] == "failure":
                state = "failing"
            elif (
                r["last_status"] is None
                and last_success is None
                and r["last_failure_at"] is None
            ):
                state = "never_run"
            elif interval is not None and interval > 0 and (
                now - last_success
            ).total_seconds() > _STALE_INTERVAL_MULTIPLIER * float(interval):
                state = "stale"
            else:
                state = "ok"
            health_loops.append(
                {
                    "loop": operation_name,
                    "state": state,
                    "last_status": r["last_status"],
                    "last_success_at": last_success.isoformat()
                    if last_success
                    else None,
                    "last_failure_at": r["last_failure_at"].isoformat()
                    if r["last_failure_at"]
                    else None,
                    "consecutive_failures": consec,
                    "last_error": r["last_error"],
                    "last_result": r["last_result"],
                    "expected_interval_seconds": interval,
                }
            )
            if state != "never_run":
                sev = _SEVERITY[state]
                if sev > worst_severity:
                    worst_severity = sev
                    worst_state = state

        return {
            "now": now.isoformat(),
            "loops": loops,
            "health": {
                "overall": worst_state,
                "loops": health_loops,
            },
            "demand": {
                "active": demand["active_demand"],
                "total": demand["total_demand"],
            },
            "claims": {
                "total": claims["total"],
                "last_24h": claims["recent"],
            },
            "fill_last_hour": fill_recent,
            "production_24h": {
                "predictions": production["predictions_24h"],
                "findings": production["findings_24h"],
            },
        }

    return router

File: ui/src/lib/system.ts
import type { StalenessTier } from "./age";
import { formatAge } from "./age";
import { authedGetJson } from "./auth";

export interface LoopStatus {
  loop: string;
  last_activity: string | null;
  age_seconds: number | null;
  never_run: boolean;
}

export interface SystemStatus {
  now: string;
  loops: LoopStatus[];
  demand: { active: number; total: number };
  fill_last_hour: Record<string, number>;
  production_24h: { predictions: number; findings: number };
}

// /system/status is auth-gated (resolve_audience_from_request -> 401 with no
// token), so this goes through authedGetJson: an absent token throws
// AuthRequiredError rather than a silent anonymous fetch that would always 401.
export const getSystemStatus = (): Promise<SystemStatus> =>
  authedGetJson<SystemStatus>("/system/status");

export type LoopCadence = "scheduled" | "on_demand";

// A loop's cadence decides whether staleness is a *health* signal or just a
// quiet period. Scheduled loops write on a tick, so silence means the engine
// stopped; on_demand loops write only when there is something to say, so
// silence is a legitimate state (a quiet week of findings is the system
// working, per the conviction gate). Grading an on_demand loop on the same
// staleness scale as a scheduled one would manufacture a false alarm exactly
// where the product is designed to stay quiet.
export const LOOP_CADENCE: Record<string, LoopCadence> = {
  prediction: "scheduled",
  fill: "scheduled",
  finding: "on_demand",
  demand: "on_demand",
  claim_ingest: "on_demand",
};

export function loopCadence(name: string): LoopCadence {
  return LOOP_CADENCE[name] ?? "on_demand";
}

const MINUTE = 60;
const HOUR = 60 * MINUTE;

// Operational thresholds for scheduled loops. These are glance-grade, not SLAs:
// a prediction loop that last wrote over an hour ago has missed several ticks
// and warrants attention; one idle for six hours is almost certainly stuck.
// never_run is its own state (a fresh deployment), distinct from a dead loop,
// and must never be graded as "dead" -- that would conflate "hasn't started"
// with "broke".
export function scheduledLoopTier(
  ageSeconds: number | null,
  neverRun: boolean,
): StalenessTier {
  if (neverRun) return "unknown";
  if (ageSeconds === null || Number.isNaN(ageSeconds)) return "unknown";
  if (ageSeconds < 0) return "fresh";
  if (ageSeconds < 2 * MINUTE) return "fresh";
  if (ageSeconds < 15 * MINUTE) return "recent";
  if (ageSeconds < HOUR) return "aging";
  if (ageSeconds < 6 * HOUR) return "stale";
  return "dead";
}

// The worst health among scheduled loops only -- the single "is the engine
// alive" reading for the compact rail. On_demand loops are excluded so a
// healthy silence in findings/demand cannot raise the headline tier.
const TIER_RANK: Record<StalenessTier, number> = {
  fresh: 0,
  recent: 1,
  aging: 2,
  stale: 3,
  dead: 4,
  unknown: 5,
};

export function worstScheduledTier(loops: LoopStatus[]): StalenessTier {
  const scheduled = loops.filter((l) => loopCadence(l.loop) === "scheduled");
  if (scheduled.length === 0) return "unknown";
  let worst: StalenessTier = "fresh";
  for (const l of scheduled) {
    const tier = scheduledLoopTier(l.age_seconds, l.never_run);
    if (TIER_RANK[tier] > TIER_RANK[worst]) worst = tier;
  }
  return worst;
}

export type FillOutcomeClass = "good" | "blocked" | "failed" | "neutral";

// fill_outcome enum is fixed at ('filled', 'unfillable', 'error') in
// 001_core_schema.sql. `error` is the actionable one (provider auth/network/
// schema failure); `unfillable` is demand the catalog genuinely cannot satisfy,
// which is a real signal but not a crash. Any future outcome renders neutral
// rather than being silently miscategorized.
export function fillOutcomeClass(outcome: string): FillOutcomeClass {
  if (outcome === "filled") return "good";
  if (outcome === "unfillable") return "blocked";
  if (outcome === "error") return "failed";
  return "neutral";
}

// never_run is rendered as a distinct label rather than an age, because
// formatAge(null) would say "no data" -- which reads as a failure to fetch the
// timestamp, not the truthful "this loop has never produced output yet".
export type EngineStatusWord =
  | "nominal"
  | "degraded"
  | "stalled"
  | "down"
  | "standby";

// The single headline word the rail shows for overall engine health. It maps
// the worst scheduled-loop tier to a state a glance can parse: nominal needs no
// action, degraded means ticks are slipping, stalled means a loop has missed
// many cycles, down means almost certainly stopped. "standby" is the fresh
// deployment (never_run) -- distinct from down so first install does not read
// as an outage.
export function engineStatusWord(tier: StalenessTier): EngineStatusWord {
  switch (tier) {
    case "fresh":
    case "recent":
      return "nominal";
    case "aging":
      return "degraded";
    case "stale":
      return "stalled";
    case "dead":
      return "down";
    case "unknown":
    default:
      return "standby";
  }
}

// never_run is rendered as a distinct label rather than an age, because
// formatAge(null) would say "no data" -- which reads as a failure to fetch the
// timestamp, not the truthful "this loop has never produced output yet".
export function loopAgeLabel(loop: LoopStatus): string {
  if (loop.never_run) return "never run";
  return formatAge(loop.age_seconds);
}
